File: src/test/scala/org/rucksac/parser/css/PerformanceTest.java
package org.rucksac.parser.css;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.junit.Before;
import org.junit.Test;
import org.w3c.dom.Attr;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

/**
 * @author Oliver Becker
 * @since 31.08.12
 */
public class PerformanceTest {

    private Document document;


    private Element createElement(String name, String id, String styleClass, Attr attr) {
        Element result = this.document.createElement(name);
        if (id != null) {
            result.getAttributes().setNamedItem(createAttribute("id", id));
        }
        if (styleClass != null) {
            result.getAttributes().setNamedItem(createAttribute("class", styleClass));
        }
        if (attr != null) {
            result.getAttributes().setNamedItem(attr);
        }
        return result;
    }


    private Attr createAttribute(String name, String value) {
        Attr attribute = this.document.createAttribute(name);
        attribute.setValue(value);
        return attribute;
    }

    @Before
    public void setup() throws ParserConfigurationException {
        this.document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
        Element root = createElement("root", "doc", null, null);
        for (int i = 0; i < 400; i++) {
            Element child = createElement("first", null, "eins uno un one", createAttribute("name", "bim"));
            root.appendChild(child);
            for (int j = 0; j < 1000; j++) {
                child.appendChild(createElement("second", null, "zwei dos deux two", createAttribute("name", "bam")));
                child.appendChild(createElement("third", null, "drei tres trois three", createAttribute("name", "bum")));
            }
        }
        this.document.appendChild(root);
    }

    @Test
    public void testMatch() throws Exception {
        new Query<Node>("root#doc > first.one.uno[name='bim'] *[name]").filter(this.document.getDocumentElement());
    }

}

File: src/test/scala/org/rucksac/parser/css/PerformanceTest.scala
package org.rucksac.parser.css

import org.w3c.dom.{Attr, Document}
import org.junit.{Ignore, Test, Before}
import org.junit.Assert._
import javax.xml.parsers.DocumentBuilderFactory
import org.rucksac.matcher.NodeMatcherRegistry

/**
 * @author Andreas Kuhrwahl
 * @since 01.09.12
 */
@Ignore
class PerformanceTest {

    private var document: Document = null

    private def createElement(name: String, id: String, styleClass: String, attr: Attr) = {
        val result = document.createElement(name)
        if (id != null) {
            result.getAttributes.setNamedItem(createAttribute("id", id))
        }
        if (styleClass != null) {
            result.getAttributes.setNamedItem(createAttribute("class", styleClass))
        }
        if (attr != null) {
            result.getAttributes.setNamedItem(attr)
        }
        result
    }

    private def createAttribute(name: String, value: String) = {
        val attr = document.createAttribute(name)
        attr.setValue(value)
        attr
    }

    val loop1 = 200
    val loop2 = 100

    @Before
    def setup() {
        NodeMatcherRegistry.all()

        document = DocumentBuilderFactory.newInstance.newDocumentBuilder.newDocument
        val root = createElement("root", "doc", null, null)
        (0 until loop1) foreach {
            i =>
                val child = createElement("first", null, "eins uno un one", createAttribute("name", "bim"))
                root.appendChild(child)
                (0 until loop2) foreach {
                    j =>
                        child.appendChild(
                            createElement("second", null, "zwei dos deux two", createAttribute("name", "bam")))
                        child.appendChild(
                            createElement("third", null, "drei tres trois three", createAttribute("name", "bum")))
                }
        }
        document.appendChild(root)
    }

    @Test
    def testPredicate() {
        executeTest("root#doc > first.one.uno[name='bim'] *[name]")
    }

    @Test
    def testFilter() {
        executeTest("root#doc:eq(0) > first.one.uno[name='bim']:gt(-1) *[name]:gt(-1)")
    }

    private def executeTest(expression: String) {
        System.gc()
        val start = System.currentTimeMillis()
        val result = $(expression, document)
        println("Query took " + (System.currentTimeMillis() - start)/1000.0 + " seconds")
        assertEquals(loop1 * loop2 * 2, result.size)
    }

}
